File: src/preprocessing/stanseg.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# @file stanseg.py
# @brief Wrapper for the Stanford Segmenter with our customized handling for hashtags and wa
#
# @author Eric Joanis

# We encourage use of Python 3 features such as the print function.
from __future__ import print_function, unicode_literals, division, absolute_import

import os
import sys
import codecs
import re
from argparse import ArgumentParser
from portage_utils import *
from subprocess import *

def get_args():
   """Command line argument processing."""

   usage="stanseg.py [options] [infile [outfile]]"
   help="""
   Takes OSPL input, pass it through the Stanford Segmenter with some
   additional conditional processing.
   """

   # Use the argparse module, not the deprecated optparse module.
   parser = ArgumentParser(usage=usage, description=help, add_help=False)

   # Use our standard help, verbose and debug support.
   parser.add_argument("-h", "-help", "--help", action=HelpAction)
   parser.add_argument("-v", "--verbose", action=VerboseAction)
   parser.add_argument("-d", "--debug", action=DebugAction)

   parser.add_argument("-m", dest="xmlishifyHashtags", action='store_true', default=False,
                       help="xmlishify hashtags. [%(default)s]")

   parser.add_argument("-w", dest="removeWaw", action='store_true', default=False,
                       help="Remove beginning of sentence's w+. [%(default)s]")

   # The following use the portage_utils version of open to open files.
   parser.add_argument("infile", nargs='?', type=open, default=sys.stdin,
                       help="input file [sys.stdin]")
   parser.add_argument("outfile", nargs='?', type=lambda f: open(f,'w'),
                       default=sys.stdout,
                       help="output file [sys.stdout]")

   cmd_args = parser.parse_args()

   return cmd_args

class RequestPostprocessor():
   def __init__(self, removeWaw=False, xmlishifyHashtags=False, unescapeHandles=False):
      self.removeWaw = removeWaw
      self.xmlishifyHashtags = xmlishifyHashtags
      self.unescapeHandles = unescapeHandles
      self.re_escaped_handle = re.compile(ur'^TWITTERHANDLEZA[a-zA-Z]+$')
      self.re_handle_reverse_dict = {
         'A': '@', 'U': '_', 'Z': 'Z',
         'a': '0', 'b': '1', 'c': '2', 'd': '3', 'e': '4',
         'f': '5', 'g': '6', 'h': '7', 'i': '8', 'j': '9',
      }

   def _escapeXMLChars(self, sentence):
      # < hashtag > -> <hashtag> and </ hashtag > -> </hashtag> since Stan Seg toks them
      sentence = re.sub("(^| )<(/|) hashtag >($| )", "\g<1><\g<2>hashtag>\g<3>", sentence)
      # We need to run this RE sub twice because the first misses a < hashtag >
      # immediately after a </ hashtag >
      sentence = re.sub("(^| )<(/|) hashtag >($| )", "\g<1><\g<2>hashtag>\g<3>", sentence)

      tokens = sentence.split()
      for i, tok in enumerate(tokens):
         if tok not in ("<hashtag>", "</hashtag>"):
            tok = re.sub("&(?![a-zA-Z]+;)","&amp;",tok)
            tok = re.sub("<","&lt;",tok)
            tok = re.sub(">","&gt;",tok)
            tokens[i] = tok
      return ' '.join(tokens)

   def _unescapeTwitterHandle(self, sentence):
      tokens = sentence.split()
      prefix_len = len("TWITTERHANDLE")
      for i, tok in enumerate(tokens):
         if self.re_escaped_handle.match(tok):
            #print("unescape {}".format(tok), file=sys.stderr)
            tokens[i] = re.sub('Z(.)',
               lambda m: self.re_handle_reverse_dict.get(m.group(1), m.group(1)),
               tok[prefix_len:])
      return ' '.join(tokens)

   def _removeWaw(self, sentence):
      return re.sub(u'^و\+ ', '', sentence)

   def __call__(self, sentence):
      if self.xmlishifyHashtags:
         sentence = self._escapeXMLChars(sentence)
      if self.removeWaw:
         sentence = self._removeWaw(sentence)
      if self.unescapeHandles:
         sentence = self._unescapeTwitterHandle(sentence)
      return sentence

def run_normalize(infile):
   """
   Pass the input through normalize-unicode.pl ar since the Stanford Segmenter
   does not recognize characters in their presentation form, only in their
   canonical form.
   """
   norm_cmd = "normalize-unicode.pl ar"
   p = Popen(norm_cmd, shell=True, stdin=infile, stdout=PIPE).stdout
   return p

def run_prepro(infile, xmlishifyHashtags):
   """
   Run the preprocessing in a separate script before calling the stanford
   segmenter itself. The code might have been placed in a child process in this
   script by using os.fork(), but writing a separate script is simpler (and
   lazier, I know...)
   """
   pre_cmd = "stanseg-pre.py"
   if xmlishifyHashtags:
      pre_cmd += " -m"
   p = Popen(pre_cmd, shell=True, stdin=infile, stdout=PIPE).stdout
   return p

def run_stan_seg(infile):
   """Run the Stanford Segmenter itself"""
   stanseg_home = os.environ.get('STANFORD_SEGMENTER_HOME', None)
   stanseg_classifier = "arabic-segmenter-atb+bn+arztrain.ser.gz"
   stanseg_cmd = "java -mx4g " + \
      "edu.stanford.nlp.international.arabic.process.ArabicSegmenter " + \
      "-loadClassifier " + stanseg_home + "/data/" + stanseg_classifier + \
      " -prefixMarker + -suffixMarker + -domain arz -nthreads 1"
   p = Popen(stanseg_cmd, shell=True, stdin=infile, stdout=PIPE).stdout
   #print(p)
   #print(p[0])
   return p

def main():
   os.environ['PORTAGE_INTERNAL_CALL'] = '1';   # add this if needed

   print("DEPRECATED WARNING: stanseg.py was the first attempt to wrap the Stanford Segmenter and is somewhat buggy. It's here for development and documentation purposes only. Use stanseg.pl instead.\n", file=sys.stderr)

   cmd_args = get_args()

   pipe1 = run_normalize(cmd_args.infile)
   pipe2 = run_prepro(pipe1, cmd_args.xmlishifyHashtags)
   pipe3 = codecs.getreader('utf-8')(run_stan_seg(pipe2))

   outfile = codecs.getwriter('utf-8')(cmd_args.outfile)

   post = RequestPostprocessor(cmd_args.removeWaw, cmd_args.xmlishifyHashtags,
      unescapeHandles=True)
   for line in pipe3:
      print(post(line.rstrip()), file=outfile)


if __name__ == '__main__':
   main()

File: src/preprocessing/stanseg.pl
#!/usr/bin/env perl

# @file stanseg.pl
# @brief Wrapper for the Stanford Segmenter
#
# @author Eric Joanis
#
# Traitement multilingue de textes / Multilingual Text Processing
# Centre de recherche en technologies numériques / Digital Technologies Research Centre
# Conseil national de recherches Canada / National Research Council Canada
# Copyright 2019, Sa Majeste la Reine du Chef du Canada /
# Copyright 2019, Her Majesty in Right of Canada

use utf8;
use strict;
use warnings;
use 5.014;

BEGIN {
   # If this script is run from within src/ rather than being properly
   # installed, we need to add utils/ to the Perl library include path (@INC).
   if ( $0 !~ m#/bin/[^/]*$# ) {
      my $bin_path = $0;
      $bin_path =~ s#/[^/]*$##;
      unshift @INC, "$bin_path/../utils", $bin_path;
   }
}
use portage_utils;
printCopyright(2019);
$ENV{PORTAGE_INTERNAL_CALL} = 1;

use ULexiTools;
setTokenizationLang("en"); # for tokenizing tags themselves, not real text

sub usage {
   local $, = "\n";
   print STDERR @_, "";
   $0 =~ s#.*/##;
   print STDERR "
Usage: $0 [-m] [-w] [infile [outfile]]

  Wrapper around the Stanford Segmenter that processes non-arabic text and
  numbers in a way that Portage needs.

  Input should be one-sentence-per-line. (Use sentsplit_plugin ar if necessary.)
  Output will be line-aligned to the input.

Options:

  -m            xmlishify hashtags
  -w(aw)        remove waw prefix at the beginning of the sentence
  -t(hreads) T  Run the Stanford Segmenter with T threads and T GB of RAM

";
   exit @_ ? 1 : 0;
}



use Getopt::Long;
Getopt::Long::Configure("no_ignore_case");
# Note to programmer: Getopt::Long automatically accepts unambiguous
# abbreviations for all options.
my $verbose = 1;
GetOptions(
   help        => sub { usage },
   debug       => \my $debug,

   m           => \my $xmlishify_hashtags,
   waw         => \my $remove_waw,
   "threads=i" => \my $threads,
) or usage "Error: Invalid option(s).";

$xmlishify_hashtags = 0 unless defined $xmlishify_hashtags;
$remove_waw = 0 unless defined $remove_waw;
$threads = 1 unless defined $threads;

my $in  = shift || "-";
my $out = shift || "-";

0 == @ARGV or usage "Error: Superfluous argument(s): @ARGV";

open IN, "<$in" or die "Cannot open input file $in: $!\n";
open OUT, ">$out" or die "Cannot open output file $out: $!\n";
binmode IN, ":encoding(UTF-8)";
binmode OUT, ":encoding(UTF-8)";
binmode STDERR, ":encoding(UTF-8)";


use File::Temp qw/tempfile/;
use File::Spec;
my ($ar_fh, $ar_filename) =
   tempfile("stanseg-tmp-ar.XXXX", DIR=>File::Spec->tmpdir(), UNLINK=>1);
binmode $ar_fh, ":encoding(UTF-8)";
my ($non_ar_fh, $non_ar_filename) =
   tempfile("stanseg-tmp-non-ar.XXXX", DIR=>File::Spec->tmpdir(), UNLINK=>1);
binmode $non_ar_fh, ":encoding(UTF-8)";
my ($tok_non_ar_fh, $tok_non_ar_filename) =
   tempfile("stanseg-tmp-non-ar-tok.XXXX", DIR=>File::Spec->tmpdir(), UNLINK=>1);
binmode $tok_non_ar_fh, ":encoding(UTF-8)";


### Phase 1: Split the input into two streams: Arabic and non-Arabic

my $last_arabic_id=0;
while (<IN>) {
   chomp;

   #s/[\p{Arabic}]/A/g;
   s/__ARABIC__ID/__ESCAPE_ARABIC_ID/g; # Avoid collision with ID marker

   # transform arabic numbers into English ones
   tr/٠-٩۰-۹٪٫٬/0-90-9%.,/;

   if ($xmlishify_hashtags) {
      s/&(?![a-zA-Z]+;)/&amp;/g;
      s/</&lt;/g;
      s/>/&gt;/g;
      s|#([^ #]+)|
         my $tokenized_hashtag = $1;
         $tokenized_hashtag =~ tr/_/ /;
         " <hashtag> $tokenized_hashtag </hashtag> "
      |eg;
   }

   # Extract Arabic text to feed to the Stanford Segmenter
   s/(\p{Script_Extensions: Arabic}[\p{Script_Extensions: Arabic}\s_]+)/
      #print STDERR "ARABIC $1\n";
      ++$last_arabic_id;
      print {$ar_fh} "BEGIN " if ($-[0] != 0);
      print {$ar_fh} "$1 END\n";
      " __ARABIC__ID" . $last_arabic_id . "__ "
   /eg;

   print {$non_ar_fh} $_, "\n";
}
close IN;
close $ar_fh;
close $non_ar_fh;

if ($debug) {
   system("cat $ar_filename");
   system("cat $non_ar_filename");
   system("ls -l $ar_filename $non_ar_filename");
}


### Phase 2: Tokenize the non-Arabic stream

{
   my $lang = "en";
   my $v = 0;
   my $p = 0;
   my $ss = 0;
   my $noss = 1; # We're working on OSPL: don't split sentences
   my $notok = 0;
   my $pretok = 0;
   my $paraline = 0;
   # If we are working with xml-ish tags, preserve them.
   my $xtags = $xmlishify_hashtags;

   tokenize_file($non_ar_filename, $tok_non_ar_filename, $lang, $v, $p, $ss, $noss, $notok,
                 $pretok, $paraline, $xtags) == 0
      or die "Error: stanseg.pl encountered a fatal error running tokenizer\n";
}

$debug and system("ls -l $tok_non_ar_filename");


### Phase 3: use the Stanford Segmenter on the Arabic stream

open NON_AR_IN, "<$tok_non_ar_filename" or die "Cannot opem temporarily non-Arabic stream file $tok_non_ar_filename: $!\n";
binmode NON_AR_IN, ":encoding(UTF-8)";

my $stanseg_home = $ENV{'STANFORD_SEGMENTER_HOME'};
my $stanseg_classifier = "arabic-segmenter-atb+bn+arztrain.ser.gz";
my $stanseg_cmd = "java -mx" . $threads . "g " .
      "edu.stanford.nlp.international.arabic.process.ArabicSegmenter " .
      "-loadClassifier " . $stanseg_home . "/data/" . $stanseg_classifier .
      " -prefixMarker + -suffixMarker + -domain arz -nthreads " . $threads;
open STAN_SEG_PIPE, "normalize-unicode.pl ar < $ar_filename | $stanseg_cmd |" or die "Cannot open Stanford segmenter pipe: $!\n";
binmode STAN_SEG_PIPE, ":encoding(UTF-8)";


### Phase 4: merge the processed Arabic and non-Arabic streams

while (<NON_AR_IN>) {
   s|__ARABIC__ID([0-9]+)__|
      my $contents = <STAN_SEG_PIPE>;
      die "Error: Missing line in $ar_filename for ARABIC ID$1" if (!defined $contents);
      chomp $contents;
      $contents =~ s/^BEGIN +//;
      $contents =~ s/ +END$//;
      $contents
   |eg;
   s/__ESCAPE_ARABIC_ID/__ARABIC__ID/g; # Undo collision avoidance
   s/   */ /g;
   s/ *$//;
   s/^ *//;
   s/^و\+ // if $remove_waw;
   print OUT;
}

close STAN_SEG_PIPE or die "Error closing Stanford segmenter pipe: $!\n";

